learn/security/api-design-first-security-guide.md

```diff
@@ -0,0 +1,973 @@
+
+* Missing bounds (no `maxLength`, no numeric ranges) — leads to resource exhaustion
+* Overly permissive regex patterns — allow unwanted characters and injection payloads
+* Validating at the wrong layer — front-end only, or after DB calls
+* Accepting undeclared properties — enables mass assignment
+
+Quick fixes:
+
+```yaml
+# JSON Schema example: tighten strings and numeric bounds
+type: object
+additionalProperties: false
+properties:
+  name:
+    type: string
+    minLength: 1
+    maxLength: 100
+    pattern: "^[a-zA-Z\\s]+$"
+  age:
+    type: integer
+    minimum: 0
+    maximum: 130
+```
+
+> Redocly insight: "We block PRs that add new string fields without `maxLength`. It's the cheapest guardrail against abuse."
+
+### Attack Example: Equifax (OGNL injection via Apache Struts, 2017)
+
+Here's another one. An input validation and deserialization flaw (CVE-2017-5638) in Apache Struts allowed attackers to inject OGNL expressions via the `Content-Type` header, leading to remote code execution and data exfiltration.
+
+```mermaid
+sequenceDiagram
+    participant A as Attacker
+    participant W as WAF (misconfigured)
+    participant S as Struts App (vulnerable)
+    participant DB as Database
+
+    A->>W: HTTP request with malicious Content-Type header
+    W-->>S: Pass-through (no block)
+    A->>S: OGNL payload evaluated
+    S->>S: Remote code execution
+    S->>DB: Query PII at scale
+    DB-->>A: Exfiltrated records (PII)
+```
+
+*Sequence diagram illustrating the Equifax breach: attacker exploited Apache Struts vulnerability through OGNL injection in Content-Type headers, bypassing WAF and executing code to steal personal data.*
+
+Why this matters: Strong schema validation, input allow-lists, and patch hygiene block entire classes of injection attacks.
+
+---
+
+## Rate Limiting: Preventing DoS and Abuse with API Policies
+
+A single client, whether intentionally malicious or simply due to a bug in its code, can send a massive number of requests to an API in a short period. This can overwhelm the server, degrading performance for all other users or even causing the service to crash. Rate limiting is the primary defense against this scenario.
+
+### The Library Card Analogy
+
+> **The Library Card**: A library card allows a patron to check out a maximum of ten books per month. Once that limit is reached, the system will not allow any more checkouts until the next month begins, regardless of how many times the patron asks.
+
+**Rate limiting** applies this same principle to an API. It establishes a policy that restricts the number of requests a single client can make within a specific time window (e.g., 100 requests per minute).
+
+### Documenting Rate Limits in OpenAPI
+
+While OpenAPI 3.1 doesn't include native rate-limiting objects, extension properties (prefixed with `x-`) provide a standard mechanism. The best practice is defining custom `x-rate-limit` extensions at the operation level:
```

Comment:
This. would be against the naming conventions... should be `x-rateLimit`. Also, rate limits are usually handled more dynamic ways -- based on client contracts. Some clients may be allowed X transactions per second while others allowed Y, etc...